crates/sui-core/src/authority/authority_store.rs

```diff
@@ -1083,23 +1125,48 @@ impl AuthorityStore {
                 .collect()
         };
 
-        // We record any received or deleted objects since they could be pruned.
-        let markers_to_place = received_objects
-            .iter()
-            .map(|(object_id, version, _)| {
+        // We record any received or deleted objects since they could be pruned, and smear shared
+        // object deletions in the marker table. For deleted entries in the marker table we need to
+        // make sure we don't accidentally overwite entries.
+        let markers_to_place = {
+            let received = received_objects.iter().map(|(object_id, version, _)| {
                 (
                     (epoch_id, ObjectKey(*object_id, *version)),
                     MarkerValue::Received,
                 )
-            })
-            .chain(deleted.into_iter().map(|(object_id, version)| {
+            });
+
+            let deleted = deleted.into_iter().map(|(object_id, version)| {
+                let object_key = (epoch_id, ObjectKey(object_id, version));
+                if input_objects
+                    .get(&object_id)
+                    .is_some_and(|object| object.is_shared())
+                {
+                    (
+                        object_key,
+                        MarkerValue::SharedDeleted(*transaction.digest()),
+                    )
+                } else {
+                    (object_key, MarkerValue::OwnedDeleted)
+                }
+            });
+
+            // We "smear" shared deleted objects in the marker table to allow for proper sequencing
+            // of transactions that are submitted after the deletion of the shared object.
+            // NB: that we do _not_ smear shared objects that were taken immutably in the
+            // transaction.
+            let smeared_objects = effects.deleted_mutably_accessed_shared_objects();
```

Comment:
I feel like the two places where we need to call deleted_mutably_accessed_shared_objects all have a lot of complicated code, that there may be opportunities to simplify them if we abstract more logic into effects. But it's not immediately obvious what could be done. Something to look into.

Reply:
Let me think on this a bit, but at least from looking at the two usages of this I don't see any way to abstract it away more than it currently is at the moment. Definitely something worth pondering though...